VSR: Nack during DVC-phase of view-change, not repair-phase

sentientwaffle opened the pull request:
Move the nack-prepare phase into the DVC step.
How this works:
- Use `dvc.header.context` as a "missing" bitset to indicate nack-able (or situationally-nackable) messages (see DVCQuorum's updated docs for detail).
- Don't send "fault" entries in the DVC headers anymore — send the real header if available, and a blank otherwise.
- The new primary uses the combination of the "missing" bitset with the headers/blanks to nack+truncate before the repair phase of the view change even begins.

Note that this design means that `pipeline_prepare_queue_max` now has a hard limit of `@bitSizeOf(u128) - 1`.
We won't use a value that high even in production.

## Pre-merge checklist

Performance:


* [x] I am very sure this PR could not affect performance.

sentientwaffle committed:
VSR: Nack prepares during DVC phase

Move the nack-prepare phase into the DVC step.
How this works:
- Use `dvc.header.context` as a "missing" bitset to indicate nack-able (or situationally-nackable) messages (see DVCQuorum's updated docs for detail).
- Don't send "fault" entries in the DVC headers anymore — send the real header if available, and a blank otherwise.
- The new primary uses the combination of the "missing" bitset with the headers/blanks to nack+truncate before the repair phase of the view change even begins.

(This commit leaves the `nack_prepare` protocol intact, but it is no longer used.)

Note that this design means that `pipeline_prepare_queue_max` now has a hard limit of `@bitSizeOf(u128) - 1`.
We won't use a value that high even in production.

sentientwaffle committed:
VSR: Remove nack-prepare protocol

This part of CTRL is part of the view change now, so the separate protocol is unnecessary.

matklad commented on src/vsr.zig:
```diff
@@ -1076,6 +1062,8 @@ pub fn quorums(replica_count: u8) struct {
     assert(quorum_view_change + quorum_replication > replica_count);
 
+    // We need to have enough nacks to guarantee that `quorum_replication` was not reached,
+    // because if the replication quorum was reached, then it may have been committed.
```
```suggestion
    // because if the replication quorum was reached, then it is committed.
```

It is impossible for an entry to not be committed if it collected replication quorum of prepare-oks, right?

sentientwaffle replied:
There might be ops preceding it in the pipeline which do not have a replication quorum of prepare-oks.
Or alternately, it the prepare oks might have been lost, and the op is "logically" committed (i.e. it will eventually be committed), but no replica has physically executed it in that state machine yet.

matklad replied:
Wrong: it might be truncated if some preceding entry is not committed.

matklad commented on src/vsr/replica.zig:
```diff
@@ -6541,36 +6336,51 @@ pub fn ReplicaType(
 ///   (Otherwise the replica would be in status=recovering_head and unable to participate).
 ///
-/// Across all DVCs in the quorum:
+/// Invariants (across all DVCs in the quorum):
```
Let's add invariant which was violated in that op 68 case:

> Any pipeline message which could have been committed is included in some canonical message

matklad commented on src/vsr/replica.zig:
```diff
@@ -6541,36 +6336,51 @@ pub fn ReplicaType(
+///
+/// Each header in the DVC body is one of:
+///
+/// | Blank | Missing |  Nack | Description
```
Some slight misgivings about terminology here, as "Missing" I think doesn't capture what we want here (that replica guarantees that it didn't send prepare-ok).

It feels like the right name for missing is probably "nack", as that's replcia acknowledging not sending  the ok out. 

The third column is an observation that we can derive nacks for different non-blank messages.

matklad commented on src/vsr/replica.zig:
```diff
@@ -3092,6 +2932,29 @@ pub fn ReplicaType(
+                        if (!self.journal.faulty.bit(slot)) {
+                            missing.set(i);
+                        }
+                    }
```
Having  bit of a hard time convincing myself that the case analysis is correct here. In particular, in 

```zig
else {
    missing.set(i);
}
```

branch, shouldn't we check that the faulty bit?

matklad replied:
Ah, it shouldn't, because, even if it faulty, we don't care, because we know it to be different. Maybe it makes sense to try to flatten this out a bit?

```zig
const slot = self.journal.slot_for_op(header.op);
const journal_header = self.journal.header_for_op(header.op);
const dirty = self.journal.dirty.bit(slot);
const faulty = self.journal.faulty.bit(slot);

// Case 1: we have this prepare in memory, but haven't persisted it to disk yet.
if (journal_header != null and journal_header.?.checksum == header.checksum and dirty and !faulty) {
    missing.set(i);
}
// Case 2: we have a _different_ prepare, safe to nack even if it is faulty.
if (journal_header != null and journal_header.?.checksum != header.checksum) {
    missing.set(i);
}
// Case 3: we don't have a prepare all, and that's not due to a fault.
if (journal_header == null and !faulty) {
    missing.set(i);
}
```

matklad replied:
Also, the difference between `header_for_op/header_with_op` is subtle. I woneder if we should use `header_for_slot` instead?

matklad replied:
Also, for case2, can we assert that that's an older prepare? That it is either for a different wrap, or has a smaller view? (not sure if both are possible here)

sentientwaffle replied:
> Also, for case2, can we assert that that's an older prepare? That it is either for a different wrap, or has a smaller view? (not sure if both are possible here)

Tricky case. I think we can't assert this, since successive view-changes can swing the version of an op forward-and-back, and crash/recovery restores the journal state. So even though the header in the view_headers is logically newer — i.e. part of a higher log_view — it might have originated in an older view.

matklad commented on src/vsr/replica.zig:
```diff
@@ -6827,40 +6656,47 @@ const DVCQuorum = struct {
+                    .valid => found != null and found.?.checksum != header.checksum,
+                    // The replica's prepare is faulty, with an unknown header.
+                    .blank => false,
+                });
```
Doesn't look to pretty, maybe an explicit if would read better here?

matklad commented on src/vsr/replica.zig:
```diff
@@ -6802,23 +6619,35 @@ const DVCQuorum = struct {
+                assert(header.op == op);
+
+                if (vsr.Headers.dvc_header_type(header) == .valid) break header;
+            } else null;
```
So here we only check that we have a header for an op, not necessary the body, right? can this create an availability issue?

That is, can we be in a situation such that:

* if we looked at DVCs from the whole cluster, we would have truncated the op
* but, looking at the DVC for the quorum we've received, we don't _nack_ this header, so it gets to be a new head
* so we send an SV with a message which should have been nacked? 

Should we also perhaps check that the `found` header is not `missing`?

matklad replied:
Yeah, I think it is an issue under a selective partition, when only some of the messages get through, and that just checking for "is not `missing`" is not going to help. 

Step 1, R0 is the primary, it prepares 1, 2, and replicates 2, but not 1

```
R0 1 2
R1   2
R2   2
```

Step 2, view change, R1 receives DVC from itself and R0. The new head is 2. 1 is found, and is not nacked (we need R2's for that). So R1 sends `SV 1 2`.

Step 3, R0 gets partitioned in such a way that its DVCs go through, but not repairs. So any subsequent view change where R0 participates gets stuck, because it doesnt' truncate  1, but can't repair it either. 

:thinking: it seems like we need that nasty "head can go backwards" bit if we want to make this 100% bullet-proof.

matklad replied:
I guess it can be less weird if `R0` crashes and corrupts `1`. Then any view change where it participates becomes stuck. Although _that_ case I think we can fix with checking for !missing.

sentientwaffle replied:
I agree we need to check that `found` is not `missing`. (I'm thinking the DVC quorum will keep waiting for DVCs if needed).

> Step 3, R0 gets partitioned in such a way that its DVCs go through, but not repairs. So any subsequent view change where R0 participates gets stuck, because it doesnt' truncate 1, but can't repair it either.

In step 2, R1 sent `SV 1 2` — if R1 sent an SV, that means that it finished repair, so by the end of step 2, the state is

```
R0 1 2
R1 1 2
R2   2
```

But I think your example applies if R1 didn't finish repair (and didn't send a SV), since its state would still be

```
R0 1 2
R1   2
R2   2
```

In that case, a view-change with R1/R2 would truncate 1,2.
But any view change involving R0 would require a repair of 1.

But isn't this outside our fault model though? (`(f+1)/(2f+1)` faults)
I guess we could artificially delay completion of the DVC quorum to allow all non-replicas a chance to register their nacks :thinking:

matklad replied:
>But isn't this outside our fault model though? ((f+1)/(2f+1) faults)

I am not sure. If R0 is thoroughly dead, that's definitely inside our fault model (1/3 is unavailable) and works. If R0 is more available than fully dead, it feels like it shouldn't make the matters worse. I think that, if we have a valid quorum, it should be able to make progress even if other nodes behave in arbitrary non-byzantine ways. 

>I guess we could artificially delay completion of the DVC quorum to allow all non-replicas a chance to register their nacks thinking

I guess we might actually get that out-of-the-box if we wait for !missing and found?

sentientwaffle replied:
> I guess we might actually get that out-of-the-box if we wait for !missing and found?

Right, the only issue left is if R0 is partially available — participating in DVCs, but not helping with repair.
But then again, maybe that would be ok too — eventually the view-change would cycle over to R0 as primary, and it already _has_ op 1.

matklad replied:
>eventually the view-change would cycle over to R0 as primary, and it already has op 1.

Right, if it's not corrupted, we'll progress, and if it _is_ corrupted, !missing would cause us to wait for full DVC set&truncate. Sounds good!

sentientwaffle committed:
VSR: Require copy of each pipeline header before view-start

Fix an availability issue, in which a view may start before any replica can finish repair.
By waiting for more DVCs, we can either truncate or repair the op.

sentientwaffle committed:
VSR: Simplify start_view_into_do_view_change

`start_view_into_do_view_change` is always called with a maxed-out SV, so we don't need to worry that there are fewer SV headers.

sentientwaffle committed:
VSR: Rename DVC's "missing" to "nack"
